src/libappimage/libappimage.cpp

```diff
@@ -0,0 +1,364 @@
+
+    try {
+        AppImage appImage(path);
+        return static_cast<utype>(appImage.getFormat());
+    } catch (const std::runtime_error& err) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    return static_cast<utype>(AppImageFormat::INVALID);
+}
+
+char** appimage_list_files(const char* path) {
+    char** result = nullptr;
+    try {
+        AppImage appImage(path);
+
+        std::vector<std::string> files;
+        for (auto itr = appImage.files(); itr != itr.end(); ++itr)
+            if (!(*itr).empty())
+                files.emplace_back(*itr);
+
+
+        result = static_cast<char**>(malloc(sizeof(char*) * (files.size() + 1)));
+        for (int i = 0; i < files.size(); i++)
+            result[i] = strdup(files[i].c_str());
+
+        result[files.size()] = nullptr;
+
+        return result;
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    // Create empty string list
+    result = static_cast<char**>(malloc(sizeof(char*)));
+    result[0] = nullptr;
+
+    return result;
+}
+
+void appimage_string_list_free(char** list) {
+    for (char** ptr = list; ptr != NULL && *ptr != NULL; ptr++)
+        free(*ptr);
+
+    free(list);
+}
+
+bool
+appimage_read_file_into_buffer_following_symlinks(const char* appimage_file_path, const char* file_path, char** buffer,
+                                                  unsigned long* buf_size) {
+    // Init output params
+    *buffer = nullptr;
+    *buf_size = 0;
+
+    std::string targetFile = file_path;
+    std::vector<std::string> visitedEntries;
+
+    while (!targetFile.empty()) {
+        // Find loops
+        if (std::find(visitedEntries.begin(), visitedEntries.end(), targetFile) != visitedEntries.end()) {
+            std::clog << "WARNING: Links loop found while extracting " << file_path;
+            break;
+        }
+
+        visitedEntries.emplace_back(targetFile);
+        std::string nextHoop;
+        try {
+            AppImage appImage(appimage_file_path);
+
+            for (auto itr = appImage.files(); itr != itr.end(); ++itr) {
+                if (itr.path() == targetFile) {
+                    if (itr.type() == PayloadEntryType::REGULAR) {
+                        auto data = std::vector<char>(std::istreambuf_iterator<char>(itr.read()),
+                                                      std::istreambuf_iterator<char>());
+
+                        *buffer = static_cast<char*>(malloc(sizeof(char) * data.size()));
+                        memcpy(*buffer, data.data(), data.size());
```

Comment:
Better use `memmove` instead of `memcpy`. Or go for an STL algorithm. Also, this is redundant code, similar to the string list creation, and there should be a helper function for it.

Reply:
In this case, using `memcpy` is safe and faster than `memmove`.

There are similar snippets of code all around but they are not exactly the same.

Reply:
`memmove` isn't slower than `memcpy` in general.

Re-reading this snippet, I would now recommend using `std::copy` here.

Reply:
See: https://stackoverflow.com/questions/4415910/memcpy-vs-memmove
But as we are using C++ let's go with `std::copy`, `std::move` will not be prudent here as we don't want our memory to be freed by the vector once destroyed.

Reply:
In this special case, moving is indeed not possible. Sorry for the noise. We have to copy anyway.